### PR TITLE
Fix MacOS CI

### PR DESCRIPTION
**Pull request by ddeclerck**

### CODE REVIEW

**Comment by GitMensch** [comment]
consider to either include libjson-c or copying cJSON.[ch] into the libcob folder

concerning the one failed test:

~~~
27. used_binaries.at:1088: testing check include header file ...
../../tests/used_binaries.at:1103: $COMPILE_MODULE prog.cob
../../tests/used_binaries.at:1107: $COMPILE_MODULE --include $(_return_path "$(pwd)/filec.h") -fstatic-call prog.cob
stderr:
/var/folders/3m/p59k4qdj0f17st0gn2cmj3640000gn/T/cob16960_0.c:108:31: error: too few arguments to function call, expected 2, have 1
  b_2 = f ((cob_u8_ptr)"Hello");
        ~                     ^
/Users/runner/work/gnucobol/gnucobol/_build/tests/testsuite.dir/0027/filec.h:3:21: note: 'f' declared here
COB_EXT_IMPORT void f (char *, long);
                    ^
1 error generated.
../../tests/used_binaries.at:1125: $COMPILE_MODULE --save-temps filec.c -o libfilec.$COB_MODULE_EXT
../../tests/used_binaries.at:1129: $COMPILE_MODULE --include $(_return_path "$(pwd)/filec.h") -fstatic-call prog2.cob
--- /dev/null	2024-05-13 14:05:02
+++ /Users/runner/work/gnucobol/gnucobol/_build/tests/testsuite.dir/at-groups/27/stderr	2024-05-13 14:05:04
@@ -0,0 +1,4 @@
+ld: Undefined symbols:
+  _f, referenced from:
+      _prog_ in cob17175_0-e76249.o
+clang: error: linker command failed with exit code 1 (use -v to see invocation)
../../tests/used_binaries.at:1129: exit code was 1, expected 0
../../tests/used_binaries.at:1129: $COMPILE_MODULE --include $(_return_path "$(pwd)/filec.h") -fstatic-call -L. -lfilec prog2.cob
stderr:
ld: unsupported mach-o filetype (only MH_OBJECT and MH_DYLIB can be linked) in '/Users/runner/work/gnucobol/gnucobol/_build/tests/testsuite.dir/0027/libfilec.dylib'
clang: error: linker command failed with exit code 1 (use -v to see invocation)
stdout:
../../tests/used_binaries.at:1129: exit code was 1, expected 0
27. used_binaries.at:1088: 27. check include header file (used_binaries.at:1088): FAILED (used_binaries.at:1129)
~~~

I have no clue how to fix that; my changes here seem to have fixed the Win32 (MinGW and MSVC) issues.

File: .github/workflows/macos.yml [hunk 1]
```diff
@@ -26,9 +26,9 @@ jobs:
...
       - name: Install packages
         run: |
-          brew install automake help2man texinfo bison berkeley-db@4
+          brew install automake libtool help2man texinfo bison berkeley-db@4
           opt="/usr/local/opt"; \
-          for d in automake help2man bison texinfo berkeley-db@4; do \
+          for d in automake libtool help2man bison texinfo berkeley-db@4; do \
             test -d "$opt/$d/bin" && echo "$opt/$d/bin" >> $GITHUB_PATH; \
             test -d "$opt/$d/lib" && export LDFLAGS="-L$opt/$d/lib $LDFLAGS"; \
             test -d "$opt/$d/include" && export CPPFLAGS="-I$opt/$d/include $CPPFLAGS"; \
```
Comment (GitMensch) [inline .github/workflows/macos.yml:34]: in theory pkg-config should find and setup lib and include directories correctly - you may want to add pkgconfig from brew (install and PATH) and see all the beauty of running with the correct setup (in theory); from the list above only berkeley-db would go to lib and include, so you may want to separate that out to an additional loop

**Comment by ddeclerck** [comment]
> consider to either include libjson-c or copying cJSON.[ch] into the libcob folder

Does this solve any specific issue ?


> concerning the one failed test:
> 
> ```
> 27. used_binaries.at:1088: testing check include header file ...
> ../../tests/used_binaries.at:1103: $COMPILE_MODULE prog.cob
> ../../tests/used_binaries.at:1107: $COMPILE_MODULE --include $(_return_path "$(pwd)/filec.h") -fstatic-call prog.cob
> stderr:
> /var/folders/3m/p59k4qdj0f17st0gn2cmj3640000gn/T/cob16960_0.c:108:31: error: too few arguments to function call, expected 2, have 1
>   b_2 = f ((cob_u8_ptr)"Hello");
>         ~                     ^
> /Users/runner/work/gnucobol/gnucobol/_build/tests/testsuite.dir/0027/filec.h:3:21: note: 'f' declared here
> COB_EXT_IMPORT void f (char *, long);
>                     ^
> 1 error generated.
> ../../tests/used_binaries.at:1125: $COMPILE_MODULE --save-temps filec.c -o libfilec.$COB_MODULE_EXT
> ../../tests/used_binaries.at:1129: $COMPILE_MODULE --include $(_return_path "$(pwd)/filec.h") -fstatic-call prog2.cob
> --- /dev/null	2024-05-13 14:05:02
> +++ /Users/runner/work/gnucobol/gnucobol/_build/tests/testsuite.dir/at-groups/27/stderr	2024-05-13 14:05:04
> @@ -0,0 +1,4 @@
> +ld: Undefined symbols:
> +  _f, referenced from:
> +      _prog_ in cob17175_0-e76249.o
> +clang: error: linker command failed with exit code 1 (use -v to see invocation)
> ../../tests/used_binaries.at:1129: exit code was 1, expected 0
> ../../tests/used_binaries.at:1129: $COMPILE_MODULE --include $(_return_path "$(pwd)/filec.h") -fstatic-call -L. -lfilec prog2.cob
> stderr:
> ld: unsupported mach-o filetype (only MH_OBJECT and MH_DYLIB can be linked) in '/Users/runner/work/gnucobol/gnucobol/_build/tests/testsuite.dir/0027/libfilec.dylib'
> clang: error: linker command failed with exit code 1 (use -v to see invocation)
> stdout:
> ../../tests/used_binaries.at:1129: exit code was 1, expected 0
> 27. used_binaries.at:1088: 27. check include header file (used_binaries.at:1088): FAILED (used_binaries.at:1129)
> ```
> 
> I have no clue how to fix that; my changes here seem to have fixed the Win32 (MinGW and MSVC) issues.

I believe this has to do with the bundle/loadable module VS dynamic library thing specific to MacOS.
I have no idea how to fix that either (and no will to delve into MacOS insanities - I have to preserve my own sanity).

File: .github/workflows/macos.yml (hunk 1 above)
Comment (ddeclerck) [reply .github/workflows/macos.yml]: Ah, indeed. I was just trying to (minimally) fix whatever was already there, but this is a useful improvement.

File: .github/workflows/macos.yml [hunk 2]
```diff
@@ -53,18 +49,18 @@ jobs:
           echo "NPROC=`sysctl -n hw.ncpu`" >> $GITHUB_ENV
           export TERM="vt100"
           echo "TERM=$TERM" >> $GITHUB_ENV
-
       - name: configure
         run: |
           cd _build
           ../configure --enable-cobc-internal-checks --enable-hardening --prefix /opt/cobol/gnucobol-gcos --exec-prefix /opt/cobol/gnucobol-gcos
+          sed -i '' 's/-undefined suppress//g' config.h
```
Comment (ddeclerck) [inline .github/workflows/macos.yml:56]: @GitMensch Also note this "hack" ; `-undefined suppress` has become deprecated on the latest linker shipped with MacOS, and it complains rather loudly (hence the failed tests).
I don't know if we can get rid of this option for every versions, including older ones...

**Comment by GitMensch** [comment]
Far sanity purpose: let's use `AT_XFAIL_IF ($COB_MODULE_EXT="dylib")` in this test (with some useful description in a comment up-front.

But that looks highly suspicions: `sed -i '' 's/-undefined suppress//g' config.h` what are we doing here and why?

Concerning the addition of libjson-c / cJSON-[hc]: without that we don't build and test a complete module (the JSON part).

**Comment by ddeclerck** [comment]
> Far sanity purpose: let's use `AT_XFAIL_IF ($COB_MODULE_EXT="dylib")` in this test (with some useful description in a comment up-front.

Alright.

> But that looks highly suspicions: `sed -i '' 's/-undefined suppress//g' config.h` what are we doing here and why?

See comment above.

> Concerning the addition of libjson-c / cJSON-[hc]: without that we don't build and test a complete module (the JSON part).

Yeah, I realized meanwhile and made the change ;)

File: .github/workflows/macos.yml (hunk 2 above)
Comment (GitMensch) [reply .github/workflows/macos.yml:56]: Hm, that should be adjusted in configure.ac. Roger changed that line last in 2005:

```diff
-	COB_SHARED_OPT="-dynamiclib"
+	COB_SHARED_OPT="-bundle -flat_namespace -undefined suppress"
```

Can you please check what these mean? Do we need them at all? I'd tend to changing that part to only be used with `rhapsody`  and comment out the `darwin` case.

When adjusting configure.ac: please fix my typo, you will find that before `ototype`.

Comment (ddeclerck) [reply .github/workflows/macos.yml:56]: Well, `-dynamiclib` is used to create a "real" dynamic library, while `-bundle` is supposed to create a loadable module, and my (superficial) understanding is that you can not "link" with a loadable module (they're meant to be dlopened).

Just out of curiosity, I'll test if the CI works with `-dynamiclib`.

Comment (ddeclerck) [reply .github/workflows/macos.yml]: Interestingly, everything works fine if I just remove the specific "darwin" case in `configure.ac`, which sets `COB_SHARED_OPT` to `-shared`. Though I'm not very confident about that (MacOS doc does not mention that option).

**Comment by GitMensch** [comment]
All checks pass - YAY.
As noted: please adjust the typo for "ototype" in configure.ac as well and check that in upstream toigether with a Changelog entry.
As we do change something "quite big" for darwin, it is likely a good idea to also note that in NEWS, what do you think?

**Comment by ddeclerck** [comment]
> All checks pass - YAY. As noted: please adjust the typo for "ototype" in configure.ac as well and check that in upstream together with a Changelog entry. As we do change something "quite big" for darwin, it is likely a good idea to also note that in NEWS, what do you think?

Changes done. I also added a NEWS entry, because this is indeed an important change.

> After doing the configure.ac related changes upstream, this PR needs to be rebased (or adjusted to only change the mirror-only workflow definitions).

I'll do that in the morning.

**Review by GitMensch — APPROVED** [review]

**Pull request merged**